trace/trace.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trace
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"k8s.io/klog"
+)
+
+type traceStep struct {
+	stepTime time.Time
+	msg      string
+}
+
+// Trace keeps track of a set of "steps" and allows us to log a specific
+// step if it took longer than its share of the total allowed time
+type Trace struct {
+	name      string
+	startTime time.Time
+	steps     []traceStep
+}
+
+// New creates a Trace with the specified name
+func New(name string) *Trace {
+	return &Trace{name, time.Now(), nil}
+}
+
+// Step adds a new step with a specific message
+func (t *Trace) Step(msg string) {
+	if t.steps == nil {
+		// traces almost always have less than 6 steps, do this to avoid more than a single allocation
+		t.steps = make([]traceStep, 0, 6)
+	}
+	t.steps = append(t.steps, traceStep{time.Now(), msg})
+}
+
+// Log is used to dump all the steps in the Trace
+func (t *Trace) Log() {
+	// an explicit logging request should dump all the steps out at the higher level
+	t.logWithStepThreshold(0)
+}
+
+func (t *Trace) logWithStepThreshold(stepThreshold time.Duration) {
+	var buffer bytes.Buffer
+	tracenum := rand.Int31()
+	endTime := time.Now()
+
+	totalTime := endTime.Sub(t.startTime)
+	buffer.WriteString(fmt.Sprintf("Trace[%d]: %q (started: %v) (total time: %v):\n", tracenum, t.name, t.startTime, totalTime))
+	lastStepTime := t.startTime
+	for _, step := range t.steps {
+		stepDuration := step.stepTime.Sub(lastStepTime)
+		if stepThreshold == 0 || stepDuration > stepThreshold || klog.V(4) {
+			buffer.WriteString(fmt.Sprintf("Trace[%d]: [%v] [%v] %v\n", tracenum, step.stepTime.Sub(t.startTime), stepDuration, step.msg))
```

Comment:
@DirectXMan12 This is an "interesting" use of logs - how do we think we would convert this to logr-style?

Reply:
it's not entirely clear to me why it's written to a buffer first, except perhaps performance.   If we remove that constraint, it looks fairly straightforward:

```go
log.Info("trace", "trace id", tracenum, "since start", step.stepTime.Sub(t.startTime), "step duration", stepDuration, "message", step.msg)
```

or better yet

```go
log := log.WithValues("trace id", tracenum)
log.Info("trace", "since start", step.stepTime.Sub(t.startTime), "step duration", stepDuration, "message", step.msg)
```

Individual traces can easily be reconstituted by filtering on `"id" == ${TRACE_ID}` or somesuch.

In the long term?  This is probably a good candidate for actual tracing ;-).

Reply:
It's at least mildly unclear what the motivation for this package is though, except to reduce logging throughput though -- it seems like a decent amount of this could be replaced with "normal" logging plus structured filtering.

Reply:
@DirectXMan12 this is extremely useful especially in the performance scaling tests. A variation of this helped me find the `nil panic propagation` problem ( see https://github.com/kubernetes/kubernetes/pull/72820/files )

Reply:
Amount of logs is one example. If something is happening 100 millions of times but it's only 100 times it's relatively slow, I don't want to see 100 millions of logs - I want to have 100.

Yes - once we have proper tracing, we probably should at least consider replacing those.